### DISCUSSION TITLE
Option to disable development mode request logging

### DISCUSSION THREAD

**Discussion by mopatches** [vercel/next.js#65992] [Ideas]
### Goals

The NextJS dev server automatically prints out every request's method and time. This prints a lot of spam if there are background processes polling the NextJS server.

We should add an option to turn off this default-on logging.

### Background

The change to add this default logging was added recently, in https://github.com/vercel/next.js/pull/62946.

### Proposal

Gate the logging by a nextConfig flag within the [`logging` subgroup](https://nextjs.org/docs/app/api-reference/next-config-js/logging).

Code to gate: https://github.com/vercel/next.js/blob/canary/packages/next/src/server/next-server.ts#L1153

```typescript
writeStdoutLine(
  `${method} ${loggingUrl} ${color(
    res.statusCode.toString()
   )} in ${reqDuration}ms`
)
```

Bonus points if the option is a "log the request if it takes longer than N milliseconds", instead of a boolean log or don't log.

**Comment by ntatoud** [comment]
👍🏼 Would be a great idea, I spent my evening trying to figure out how to turn these off...

**Comment by calebjacob** [comment]
I definitely agree we need a config option to turn off these development server logs. This is a real pain point for my application. The new logs create so much noise that it makes it very difficult to see important log information when I need it.

**Comment by alex-zabyshny** [comment]
Completely agree. Spent all evening trying to disable this logging. It is surprising that someone decided that everyone needs it. Many people install their own loggers using `pino`, `winston` and so on.

**Comment by luskin** [comment]
Same here, it's absolutely horrendous when a background service is pinging frequently! Would be great to see route-level logging config (i.e. disabling certain `/api` routes from logging).

**Comment by waltershewmake** [comment]
I would love to see this implemented. These logs make it impossible to see actual logs, especially when using tRPC.

**Comment by Haswell-s** [comment]
I was looking for it too, hope it gets implemented soon.
this feature conflicts with more advanced logging solutions.

**Comment by mopatches** [comment]
Looks like this is supported in canary. So when Next v15 comes out we'll be able to use the new option. :tada: 

* MR: https://github.com/vercel/next.js/pull/67590/files
* Release: https://github.com/vercel/next.js/releases/tag/v15.0.0-canary.77
* Docs: https://nextjs.org/docs/app/api-reference/next-config-js/logging

Thank you @huozhi :heart:

**Reply by the-me-0** [reply]
I mean don't get me wrong but I feel like this is not enough. It's already something, agreed, and it might help a handfull of devs, but it would be even better if we could (as mentioned by others before me) disable logs for some routes in particular. I got a /api/auth/session in my project that gets spammed and it just makes the logs difficult to read tbh.

Hope it gets implemented soon.

**Reply by huksley** [reply]
The worst thing is it does logging even if you use custom server (in development)

I am using this **hack**, it is horrible but works well.

```js
if (process.env.NODE_ENV === "development") {
      // Omit NextJS writing it's log https://github.com/vercel/next.js/discussions/65992
      const __write = process.stdout.write;
      // @ts-ignore
      process.stdout.write = (...args) => {
        if (
          typeof args[0] !== "string" ||
          !(
            args[0].startsWith(" GET /") ||
            args[0].startsWith(" POST /") ||
            args[0].startsWith(" DELETE /") ||
            args[0].startsWith(" PATCH /")
          )
        ) {
          // @ts-ignore
          __write.apply(process.stdout, args);
        }
      };
    }
```

**Reply by AuthorOfTheSurf** [reply]
This is what I opted for as well. I am unwilling to update to NextJS 15 (or whatever version this is supposed to be fixed in) just to fix the logging. None of the configuration-related solutions worked, but maybe they work fine in newer versions of NextJS

I added this snippet to a new `util/disable-next-logging.ts` file, and then imported that file in my `layout.ts` file

Test to ensure that it's working as expected, for whatever reason this approach didn't work on my first try a month or two ago, then today it worked smoothly

```ts
import "./util/disable-next-logging";
```

**Comment by huksley** [comment]
Same here. I use custom server (even in development) and I do my own request/response logging when needed. So it gets duplicated with NextJS outputting the log after the response.

**Comment by mohsen1** [comment]
I took a stab at this. Let me know if the configuration options makes sense to you https://github.com/vercel/next.js/pull/74349

**Comment by simistern** [comment]
I am struggling big time with this, I have not found anything that prevents these logs in dev mode. Any ideas are super welcome here. I have tried the ideas from above but nothing is really working for me. 

![Screenshot 2025-02-10 at 12 29 24 PM](https://github.com/user-attachments/assets/af5825a4-ad29-41fd-9aab-a1b995a74eab)
